notifications-worker.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	expo "github.com/darnfish/exponent-server-sdk-golang/sdk"
+	"github.com/dxe/alc-mobile-api/model"
+	"github.com/jmoiron/sqlx"
+)
+
+func NotificationsWorker(db *sqlx.DB) {
+	log.Println("Notifications Worker started.")
+
+	tx, err := db.Beginx()
+	if err != nil {
+		log.Println("Failed to begin database transaction.")
+		return
+	}
+
+	notifications, err := model.SelectNotificationsToSend(tx)
+	if err != nil {
+		log.Printf(err.Error())
+		tx.Rollback()
+		return
+	}
+
+	if len(notifications) == 0 {
+		log.Println("Found no queued notifications to send.")
+		tx.Rollback()
+		return
+	}
+
+	expoMessages := make([]expo.PushMessage, len(notifications))
+	for i, n := range notifications {
+		pushToken, err := expo.NewExponentPushToken(n.ExpoPushToken)
+		if err != nil {
+			log.Printf("User %v has invalid push token.", n.UserID)
+		}
+		expoMessages[i] = expo.PushMessage{
+			To:    []expo.ExponentPushToken{pushToken},
+			Title: n.Title,
+			Body:  n.Body,
+		}
+	}
+
+	client := expo.NewPushClient(&expo.ClientConfig{AccessToken: os.Getenv("EXPO_PUSH_ACCESS_TOKEN")})
+
+	expoResponses, err := client.PublishMultiple(expoMessages)
```

Comment:
I'd recommend trying to stick to database operations during any given transaction. Making external API calls like this can lead to long-lived transactions, which I don't think databases tend to like and can increase the risk of spurious transaction commit failures.

Instead, the usual way I've seen queue-y things like this get implemented is in three phases, something like:

1. First SQL transaction finds a batch of messages to send, and atomically changes all of their statuses from "available" to "leased until {{now + 5 minutes}}". (Tip: Make the "available" status actually "leased until 1970", so you don't actually need to handle multiple cases.)
2. Send all the notifications; make sure this doesn't take more 5 minutes (e.g., use context.WithDeadline or whatever, and probably add some extra slack to account for clock skew).
3. Do another SQL transaction to update all of the messages that were successfully handled as "sent"; optionally (but recommended) change all of the failed messages back to "available".

This limits the risk of duplicate messages to some hard failure happening during step 2, whereas the current design would lead to duplicate messages if any of the transaction fails. The current design also wouldn't work with multiple processes trying to send messages at the same time.

Reply:
I implemented most of this in the latest commit.